Automatic resources doc generator

sinabakh opened the pull request:
Fix #123

sinabakh commented:
Here is the proposed change to the resource registry. Now, the question is how we are going to generate docs?

aliscott commented:
LGTM - to generate the docs maybe we could have a make target that writes the notes to a markdown file in this repo. We can then create a component in the docs site to read that. What do you think?

sinabakh commented:
I'm not sure about the features we have in Gihub but I agree, we probably can generalte a markdown file for our releases and in the site repo, download the artifcacts of this project.
I will work on generating the markdown, meanwhile we can think more about the right way to do the aggregation.

sinabakh commented:
Now we can run:
```
infracost-generate-docs --input docs/templates --output docs/generated
```
@alikhajeh1 Do you think that's enough? This generates the whole `supported_resources.md`.

alikhajeh1 commented on internal/providers/terraform/aws/ecs_service.go:
```diff
@@ -11,6 +11,14 @@ import (
+func GetECSServiceRegistryItem() *schema.RegistryItem {
+	return &schema.RegistryItem{
+		Name:  "aws_ecs_service",
+		Notes: []string{"Fargate on-demand only."},
```
Only supports Fargate on-demand.

sinabakh replied:
Done

alikhajeh1 commented:
> Now we can run:
> 
> ```
> infracost-generate-docs --input docs/templates --output docs/generated
> ```
> 
> @alikhajeh1 Do you think that's enough? This generates the whole `supported_resources.md`.

Looks good. Here's how the end-to-end process could look like:
1. contributors runs that task and checks-in the supported_resources.md file into git. The idea being that master always has the latest, independently viewable, markdown file for supported resources. This file could be imported by https://www.infracost.io/docs/supported_resources/ from the frontend, meaning we never have to send a PR to the docs repo. The issue with using github artifacts is that I think they get zipped, so the docs frontend can't just import them from the browser.
2. contributors don't need to worry about git conflicts as they can always rebase/merge master into their branch and re-run- the task to create the latest docs.

what do you think?

sinabakh commented:
1. Regarding showing the files in frontend I'm not sure how it's possible.
We can setup a CI step so the docs get generated automatically in master branch, so contributors won't have to run this stage manually.

alikhajeh1 commented on cmd/infracost-generate-docs/main.go:
```diff
@@ -0,0 +1,128 @@
+package main
```
how do you feel about having this in scripts/docs-generator as that folder has other helpers? I was thinking cmd might be used for product subcommands (e.g. `infracost diff`) but I'm not sure yet. 
Also, what do you think about updating the Makefile so we can have `make docs`?

sinabakh replied:
Since this script is an acutal golang command line app, I prefered to put it in `cmd/`, on the other hand `infracost diff`  and other subcommands' logic should reside in the `infracost` cmd since the subcommand is just an arg.
https://github.com/influxdata/influxdb/tree/master/cmd is an example of using multiple cmds
Anyway, there is no silverbullet so we just need to have a standard, what you think?
Regarding the makefile I agree with you.

alikhajeh1 replied:
you're right, cmd makes more sense for this.

alikhajeh1 commented:
>     1. Regarding showing the files in frontend I'm not sure how it's possible.
>        We can setup a CI step so the docs get generated automatically in master branch, so contributors won't have to run this stage manually.

I looked into embedding or including markdown files and it doesn't seem possible/easy to do in `docusaurus` (our tools framework). So the options seem to be:
1. keep this PR as is, so we just run the docs generation command, manually copy/paste it into the docs repo when that needs updating (usually after we release a new version of infracost).
2. have supported_resources.md in the infracost repo, and link to it from inside https://www.infracost.io/docs. This is simple to do with the new command you added but it means people need to check that file into git. This seems ok but I'm not sure if conflicts will become annoying as they'll block PRs from being merged until the PR branch is rebased/master-merged.
3. update the infracost CI pipeline so it generates that file and saves it in the docs repo. This needs more investigations.

I think option 1 is ok for now, what do you think?

sinabakh commented:
Agreed, we can work on more automations later.

alikhajeh1 commented:
@sinabakh plz let me know if this ready to merge. BTW, do we need `no-color` and `version` in this cmd?

sinabakh commented:
@alikhajeh1 Removed `no-color` and `version`. I added some new github workflows, changes and ... so here is the best approach I could come up with:
* Contributor works on new resources like before
* Upon merging/pushing to master, new docs are generated and are commited to the master using github actions. This way master and its docs are like nightly builds.
* With new releases, a `docs.tar.gz` file is also added to assets.
* In the other repo, we need to create the same version tag, and in the actions we can download the docs.tar.gz and append its contents to the original site.

alikhajeh1 commented:
@sinabakh this is awesome! love it. I'll try to sort out the [other repo](https://github.com/infracost/docs) in the next few days.

alikhajeh1 commented:
@sinabakh I noticed that https://github.com/infracost/infracost/blob/master/docs/generated/supported_resources.md appears `aws_alb / aws_alb` twice

sinabakh commented:
#127
